### PR TITLE
Update origin_changed? to ignore protocol differences

### LINKED ISSUE #288: shards keeps downloading dependencies
Running ```shards install``` takes a very long time on a project with a number of dependencies, as ```shards``` keep refetching the same sources.

Here's an example from a ```lucky``` project (just an example, but could be anything really):

```
$ shards install
  Resolving dependencies                                                                                                                                          [19/9143]
  Fetching https://github.com/luckyframework/lucky.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/mosop/teeplate.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/avram.git
  Fetching https://github.com/kostya/blank.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/mosop/teeplate.git
  Fetching https://github.com/mosop/teeplate.git
  Fetching https://github.com/mosop/teeplate.git
  Fetching https://github.com/mosop/teeplate.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/will/crystal-pg.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/luckyframework/lucky_router.git
  Fetching https://github.com/luckyframework/shell-table.cr.git
  Fetching https://github.com/paulcsmith/cry.git
  Fetching https://github.com/mosop/cli.git
  Fetching https://github.com/mosop/optarg.git
  Fetching https://github.com/mosop/callback.git
  Fetching https://github.com/mosop/string_inflection.git
  Fetching https://github.com/mosop/string_inflection.git
  Fetching https://github.com/crystal-loot/exception_page.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/avram.git
  Fetching https://github.com/luckyframework/lucky_router.git
  Fetching https://github.com/luckyframework/shell-table.cr.git
  Fetching https://github.com/paulcsmith/cry.git
  Fetching https://github.com/crystal-loot/exception_page.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/luckyframework/authentic.git
  Fetching https://github.com/luckyframework/lucky.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/mosop/teeplate.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/avram.git
  Fetching https://github.com/kostya/blank.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/will/crystal-pg.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/kostya/blank.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/will/crystal-pg.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/kostya/blank.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/will/crystal-pg.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/kostya/blank.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/will/crystal-pg.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/kostya/blank.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/crystal-lang/crystal-db.git
  Fetching https://github.com/will/crystal-pg.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/luckyframework/lucky_router.git
  Fetching https://github.com/luckyframework/shell-table.cr.git
  Fetching https://github.com/paulcsmith/cry.git
  Fetching https://github.com/crystal-loot/exception_page.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/luckyframework/lucky_cli.git
  Fetching https://github.com/luckyframework/habitat.git
  Fetching https://github.com/luckyframework/wordsmith.git
  Fetching https://github.com/luckyframework/avram.git
  Fetching https://github.com/luckyframework/lucky_router.git
  Fetching https://github.com/luckyframework/shell-table.cr.git
  Fetching https://github.com/paulcsmith/cry.git
  Fetching https://github.com/crystal-loot/exception_page.git
  Fetching https://github.com/luckyframework/dexter.git
  Fetching https://github.com/luckyframework/lucky_cli.git
```

And so on and so on until eventually it finishes. But it takes *ages*.

shards 0.9.0 and crystal 0.29.0 but also experienced on older versions

### PR DESCRIPTION
**Pull request by kalafut**
I need an `insteadOf` git configuration and this issue has been really slowing things down, so I decided to make an attempt at fixing it. This is literally my first pushed Crystal code so any feedback is welcome!

Fixes #288

### CODE REVIEW

**Comment by straight-shoota** [comment]
Instead of hard coding the normalization for individual repository providers, it would probably be better to have a more generic solution which would work with *any*  provider.
I think it should suffice to extract hostname and path from both ssh+git and https URLs.

**Comment by kalafut** [comment]
I started that way but was a little concerned about the `git@...` portion, and whether that should be a distinguishing characteristic. i.e. whether `git@github.com:foo/bar` should equal `fred@github.com:foo/bar`. It sounds like that's not a concern, in which case I can definitely simplify this.

**Comment by kalafut** [comment]
@straight-shoota I've generalized this comparison. As it turns out there are actually a _lot_ of allowed git URL schemes. The regexes should take care of the common ones, and exactly matching origins will always work (i.e. current behavior).

**Comment by RX14** [comment]
Could we use `URI.parse` instead of the regex?

File: test/git_resolver_test.cr [hunk 1]
```diff
@@ -38,6 +38,28 @@ module Shards
...
+      assert r.origins_equal("misc_pattern", "misc_pattern")
+      assert r.origins_equal(nil, nil)
+
+      # mismatches
+      assert !r.origins_equal("https://gitlabZ.com/foo/bar", "https://gitlab.com/foo/bar")
+      assert !r.origins_equal("git@github.com:foo/bar2", "https://gitlab.com/foo/bar")
+      assert !r.origins_equal("http://bitbucket.org/foo", "sshZ://bitbucket.org/foo")
+      assert !r.origins_equal("misc_pattern", "misc_pattern_2")
+      assert !r.origins_equal("http://bitbucket.org/foo", nil)
+      assert !r.origins_equal(nil, "http://bitbucket.org/foo")
```
Comment (Sija) [inline test/git_resolver_test.cr]: please, use `refute` instead.

File: src/resolvers/git.cr [hunk 2]
```diff
@@ -243,7 +243,34 @@ module Shards
     end
 
     private def origin_changed?
-      (@origin_url ||= capture("git ls-remote --get-url origin").strip) != git_url
+      @origin_url ||= capture("git ls-remote --get-url origin").strip
```
Comment (Sija) [inline src/resolvers/git.cr]: IMO this should be extracted to:

```cr
def origin_url
  @origin_url ||= capture("git ls-remote --get-url origin").strip
end
```

Then, there's no need for additional `origins_equal` method (its body can be pasted in here - into `origin_changed?`).

File: src/resolvers/git.cr [hunk 3]
```diff
@@ -243,7 +243,34 @@ module Shards
...
+      match_1 = re.match(origin_1)
+      match_2 = re.match(origin_2)
+
+      return false if match_1.nil? || match_2.nil?
+
+      ["host", "path"].each do |element|
+        if match_1[element] != match_2[element]
+          return false
+        end
+      end
```
Comment (Sija) [inline src/resolvers/git.cr]: Please, no `Regex` is needed here, `URI.parse` should be more than enough.

File: test/git_resolver_test.cr (hunk 1 above)
Comment (kalafut) [reply test/git_resolver_test.cr]: 👍

File: src/resolvers/git.cr (hunk 2 above)
Comment (kalafut) [reply src/resolvers/git.cr]: Yes, I think that refactor will work fine.

**Comment by kalafut** [comment]
@Sija @RX14 Thanks for the review! `URI.parse` does not handle the SSH URIs that are in the scp style (e.g. `git@github.com:example/project.git`), so something additional was needed. The pair of regexes was compact, but I don't really like adding regexes either if I can avoid it.

I'd also missed @straight-shoota's comment about `local_path` being broken too for these patterns. As a result, I've redone things a bit to make a small git URI parser that uses `URI.parse` but can also deal with some other formats, while still returning a URI object for consistency. This It still feels little sketchy, though there are no regex's and it's probably better over all. It can be using for `origin_changed?` and `local_path`.

I need to add some other tests but with then push it up for y'all to look at.

File: src/resolvers/git.cr (hunk 3 above)
Comment (kalafut) [reply src/resolvers/git.cr]: See comment below

**Comment by kalafut** [comment]
I've substantially updated the PR based on feedback, and realizing that `local_path` is also sort of broken. No more ugly regexes either.

**Comment by straight-shoota** [comment]
The implementation of `parse_git_uri` could be much simpler:
```cr
private def parse_git_uri(raw_uri)
  uri = URI.parse(raw_uri)
  return uri if uri.absolute?
  
  URI.parse("ssh:#{raw_uri}")
end
```

**Comment by kalafut** [comment]
Using `.absolute` is a good change.  I don't see how the recommendation handles the main case of git+ssh urls, however. That would just leave everything in the path. e.g. https://play.crystal-lang.org/#/r/8grl

**Comment by straight-shoota** [comment]
Yeah, you're right.

**Comment by kalafut** [comment]
Looks like parse + `absolute` can get confused too:

```
uri = URI.parse("github.com:foo/bar") # scheme => github.com, host => nil, path => foo/bar
p uri.absolute? # => true
```

But that's only If the user (`git@...`) isn't included. While that style is accepted by git, it would seem to be a pretty unlikely real world case. I'd still be OK with `absolute` and not handling those user-less ssh URLs.

**Comment by straight-shoota** [comment]
I suppose shards should allow the exact same formats as accepted by git, so user-less scp-like format `example.com:foo/bar` should be valid. It seems like git treats opaque URLs as scp-like format, so expanding the condition to `uri.absolute? && !uri.opaque?` should be sufficient.

**Comment by kalafut** [comment]
Yep that seems to work OK.

File: src/resolvers/git.cr [hunk 4]
```diff
@@ -242,8 +242,40 @@ module Shards
       false
     end
 
-    private def origin_changed?
-      (@origin_url ||= capture("git ls-remote --get-url origin").strip) != git_url
+    def origin_url
```
Comment (ysbaddaden) [inline src/resolvers/git.cr]:
```suggestion
    private def origin_url
```
No need to be public AFAIK.

File: src/resolvers/git.cr [hunk 5]
```diff
@@ -242,8 +242,40 @@ module Shards
...
+      return true if origin_url.nil? || git_url.nil?
+
+      origin_parsed = parse_git_uri(origin_url)
+      git_parsed = parse_git_uri(git_url)
+
+      (origin_parsed.host != git_parsed.host) || (origin_parsed.path != git_parsed.path)
+    end
+
+    # Parses a URI string, with additional support for ssh+git URI schemes.
+    private def parse_git_uri(raw_uri)
```
Comment (ysbaddaden) [inline src/resolvers/git.cr]:
```suggestion
    private def parse_uri(raw_uri)
```

The `git` naming feels redundant —we're already in the `GitResolver` class.

File: src/resolvers/git.cr [hunk 6]
```diff
@@ -242,8 +242,40 @@ module Shards
...
+
+      # Otherwise, assume and attempt to parse the scp-style ssh URIs
+      host, _, path = raw_uri.partition(':')
+
+      if host.includes?('@')
+        user, _, host = host.partition('@')
+      end
+
+      # Normalize leading slash, matching URI parsing
+      if !path.starts_with?('/')
```
Comment (ysbaddaden) [inline src/resolvers/git.cr]:
```suggestion
      unless path.starts_with?('/')
```

File: test/git_resolver_test.cr [hunk 7]
```diff
@@ -38,6 +38,49 @@ module Shards
...
+        dependency["git"] = origin
+        refute library.origin_changed?, origin
+      end
+
+      # These alternatives should all trigger origin as changed
+      changed_origins = [
+        "https://github.com/foo/bar2",
+        "https://github.com/foos/bar",
+        "https://githubz.com/foo/bar",
+        "file:///github.com/foo/bar",
```
Comment (ysbaddaden) [inline test/git_resolver_test.cr]: This array is missing a ssh+git example.

File: src/resolvers/git.cr (hunk 4 above)
Comment (straight-shoota) [reply src/resolvers/git.cr]: Doesn't hurt either, though...

File: src/resolvers/git.cr (hunk 5 above)
Comment (kalafut) [reply src/resolvers/git.cr]: 👍

File: test/git_resolver_test.cr (hunk 7 above)
Comment (kalafut) [reply test/git_resolver_test.cr]: Added tests.

**Comment by kalafut** [comment]
All comments addressed.

**Comment by ysbaddaden** [comment]
Thanks!

**Pull request merged**